### ISSUE TITLE
HTTPProxy lacking ingressClassName field

### ISSUE THREAD

**Issue by andersmarkendahl** [projectcontour/contour#2566]
**Problem Description** 

As of Kubernetes 1.18 the Ingress resource have been updated with specific field 
`ingressClassName`, which deprecates the previously used annotation `kubernetes.io/ingress.class`, 
see: [Kubernetes Blog](https://kubernetes.io/blog/2020/04/02/improvements-to-the-ingress-api-in-kubernetes-1.18/).
The change comes with a new IngressClass resource as well.

HTTPproxy resource still relies on the annotation for selecting an ingress class.
From a user perspective this makes it complicated using both Ingress and HTTPproxy as it is hard to know which one to use.
Also, the HTTPproxy relying on an annotation deprecated in Kubernetes in itself causes a risk of confusion regarding HTTPproxy resource relevance.

**Suggestion**

HTTPproxy object should implement similar IngressClassName field as found in Ingress resource.
This would align the usage of both resources and will make sure HTTPproxy evolves together with Ingress.

**Comment by youngnick** [comment]
Thanks for logging this @Aoana! I agree, HTTPProxy should implement something like this. We've also got #2146 to implement this for Ingress itself.

We've been being careful with adding Ingress-V1-like functionality because it draws a hard line under the supported Kubernetes version Contour can possibly work with, and we wanted to give users time to catch up.

Adding this field to HTTPProxy implies that we'll need to start watching for IngressClass objects, which we don't currently do.

In short, I agree that we *should* do this, but there may be some other things we need to hit first.

**Comment by jpeach** [comment]
xref #2146

**Comment by xaleeks** [comment]
Just to be clear on the request here, in the latest v1.14 we delivered the ability for any Contour controller to pick up certain Ingress objects by linking through `IngressClass`, aka setting the `IngressClassName` field of Ingress objects.  Is the ask here to also leverage the exact same `IngressClass` object to match specific `HTTPProxy` resources?

I don’t think we have plans for this right @youngnick ?  Is Parking Lot 3 appropriately setting expectations?  It feels like this kind of parity is not the goal here, it feels like its muxing two different constructs as well.  Or am I misunderstanding the ask?

**Comment by youngnick** [comment]
I think the request is to port the IngressClass idea to work with HTTPProxy. There's no reason why HTTPProxy *couldn't* just reference an IngressClass _object_ instead of using a string _annotation_ like we do now.

To do this, we would need to:
- add an IngressClassName field to HTTPProxy
- ensure that the same rules apply to HTTPProxy IngressClassName as Ingress IngressClassName.

**Comment by xaleeks** [comment]
Yeah I suppose we can, filtering HTTPProxy resources by IngressClass.  I don’t think this is urgent, relative to the Gateway API work.  If you agree with it, we can leave it in parking lot 3 for now.
